okhttp/src/main/java/okhttp3/Dispatcher.java

```diff
@@ -173,6 +185,10 @@ synchronized void executed(RealCall call) {
   /** Used by {@code Call#execute} to signal completion. */
   synchronized void finished(Call call) {
     if (!runningSyncCalls.remove(call)) throw new AssertionError("Call wasn't in-flight!");
+
+    if (runningCallsCount() == 0 && idleCallback != null) {
```

Comment:
. . . my rule of `synchronized` is that you never do anything blocking in a `synchronized` block. (The FramedConnection writer is one exception to this rule.)